File: lexer/lexer.cpp
#include <bits/stdc++.h>
#include "../token/token.cpp"
#include <regex>

using namespace std;

bool isDigit(char ch);
bool isLetter(char ch);

class Lexer
{
public:
    int position;
    int readPosition;
    char ch;
    Lexer() {}
    Lexer(string input) : input(input)
    {
        position = 0;
        readPosition = 0;
        readChar();
    }
    void skipWhiteSpaces();
    void readChar();
    char peekChar();
    string readNumber();
    string readIdentifier();
    Token *nextToken();

private:
    string input;
};

void Lexer::readChar()
{
    if (readPosition >= input.size())
    {
        ch = 0;
    }
    else
    {
        ch = input[readPosition];
    }
    position = readPosition;
    readPosition += 1;
}

Token *Lexer::nextToken()
{
    Token *tok;
    skipWhiteSpaces();
    switch (ch)
    {
    case '(':
        tok = new Token(LPAREN, ch);
        break;
    case ')':
        tok = new Token(RPAREN, ch);
        break;
    case '+':
        tok = new Token(PLUS, ch);
        break;
    case '-':
        tok = new Token(MINUS, ch);
        break;
    case '<':
        tok = new Token(LT, ch);
        break;
    case '>':
        tok = new Token(GT, ch);
        break;
    case '=':
        // add equal operator??
        tok = new Token(EQ, ch);
        break;
    case ';':
        tok = new Token(SEMICOLON, ch);
        break;
    case ':':
        tok = new Token();
        if (peekChar() == '=')
        {
            tok->literal = ":=";
            tok->type = ASSINE;
            readChar();
        }
        else
        {
            // TODO: should flag a syntax error
        }
        break;
    case 0:
        cout << "end fo file";
        break;
    default:
        tok = new Token();
        if (isDigit(ch))
        {
            tok->type = NUMBER;
            tok->literal = readNumber();
        }
        else if (isLetter(ch))
        {
            tok->type = IDENT;
            tok->literal = readIdentifier();
        }
        break;
    }
    return tok;
}

void Lexer::skipWhiteSpaces()
{
    while (ch == '\r' || ch == '\t' || ch == '\n' || ch == ' ')
    {
        readChar();
    }
    return;
}

char Lexer::peekChar()
{
    if (readPosition >= input.size())
    {
        return 0;
    }
    return input[readPosition];
}

string Lexer::readNumber()
{
    int startIdx = position;
    while (isDigit(ch))
    {
        readChar();
    }
    return input.substr(startIdx, readPosition);
}
string Lexer::readIdentifier()
{
    int startIdx = position;
    while (isLetter(ch))
    {
        readChar();
    }
    return input.substr(startIdx, readPosition);
}

bool isDigit(char ch)
{
    return (ch >= '0' && ch <= '9');
}
bool isLetter(char ch)
{
    return (ch >= 'a' && ch <= 'z');
}

File: lexer/Lexer.java
package lexer;

import token.Token;
import token.TokenType;

public class Lexer {
    String input;
    int position;
    int readPosition;
    char ch;

    Lexer() {
    }

    public Lexer(String input) {
        this.input = input;
    }

    public Token nexToken() {
        readChar();
        skipWhiteSpaces();
        switch (this.ch) {
            case '{':
                skipComment();
                return nexToken();
            case '(':
                return new Token(TokenType.OPENBRACKET, ch);
            case ')':
                return new Token(TokenType.CLOSEDBRACKET, ch);
            case '*':
                return new Token(TokenType.MULOP, ch);
            case '/':
                return new Token(TokenType.DIV, ch);
            case '+':
                return new Token(TokenType.PLUS, ch);
            case '-':
                return new Token(TokenType.MINUS, ch);
            case '<':
                if (Character.toString(peekChar()).compareTo("=") == 0) {
                    readChar();
                    return new Token(TokenType.LTorEQ, "<=");
                }
                return new Token(TokenType.LT, ch);
            case '>':
                if (Character.toString(peekChar()).compareTo("=") == 0) {
                    readChar();
                    return new Token(TokenType.GTorEQ, ">=");
                }
                return new Token(TokenType.GT, ch);
            case '=':
                // add equal operator??
                return new Token(TokenType.EQ, ch);
            case ';':
                return new Token(TokenType.SEMICOLON, ch);
            case ':':
                if (Character.toString(peekChar()).compareTo("=") == 0) {
                    readChar();
                    return new Token(TokenType.ASSINE, ":=");
                } else {
                    return new Token(TokenType.SYNTAX_ERROR,
                            "syntax error worng token at character" + "[" + this.readPosition + "], we got :"
                                    + this.peekChar() + "expected :=");
                }
            case 0:
                return new Token(TokenType.EOF, '0');
            default:
                if (Character.isDigit(ch)) {
                    String num = readNumber();
                    Token tok = new Token(TokenType.NUMBER, num);
                    return tok;
                } else if ('a' <= ch && ch <= 'z') {
                    String indntfier = this.readIdnetifier();
                    Token tok = new Token(Token.lookup(indntfier), indntfier);
                    return tok;
                }
                return new Token(TokenType.SYNTAX_ERROR, ch);
        }
    }

    void skipWhiteSpaces() {
        // Pattern pattern = Pattern.compile(" |\t|\n|\r");
        // Matcher matcher = pattern.matcher(Character.toString(ch));
        // if (matcher.matches()) {
        if (this.ch == ' ' || this.ch == '\t' || this.ch == '\r' || this.ch == '\n') {
            this.readChar();
            skipWhiteSpaces();
        }
    }

    void readChar() {
        if (readPosition >= input.length()) {
            ch = 0;
            return;
        }
        ch = input.charAt(readPosition);
        this.position = this.readPosition;
        this.readPosition++;
    }

    char peekChar() {
        if (readPosition >= input.length()) {
            return 0;
        }
        return input.charAt(readPosition);
    }

    String readNumber() {
        int pos = this.position;
        while (Character.isDigit(peekChar())) {
            this.readChar();
        }
        return input.substring(pos, this.readPosition);
    }

    String readIdnetifier() {
        int pos = this.position;
        while ('a' <= peekChar() && peekChar() <= 'z') {
            this.readChar();
        }
        // TODO: make sure that it is a valid identifer
        return input.substring(pos, this.readPosition);
    }

    void skipComment() {
        while (Character.toString(peekChar()).compareTo("}") != 0) {
            readChar();
            if (Character.toString(this.ch).compareTo("{") == 0) {
                skipComment();
            }
            if (this.readPosition >= input.length()) {
                return;
            }
        }
        readChar();
    }
}
